vos/vos/vos.py

```diff
@@ -2737,3 +2563,293 @@ def get_job_status(self, url):
+            phase = VOFile(phase_url, conn, method="GET",
+                           follow_redirect=False).read().decode('utf-8')
+            # do not remove the line below. It is used for testing
+            logging.debug("Job URL: " + job_url + "/phase")
+            while phase in ['PENDING', 'QUEUED', 'EXECUTING', 'UNKNOWN']:
+                # poll the job. Sleeping time in between polls is doubling
+                # each time until it gets to 32sec
+                total_time_slept = 0
+                if sleep_time <= 32:
+                    sleep_time *= 2
+                slept = 0
+                if logger.getEffectiveLevel() == logging.INFO:
+                    while slept < sleep_time:
+                        sys.stdout.write(
+                            "\r%s %s" % (phase, roller[total_time_slept %
+                                                       len(roller)]))
+                        sys.stdout.flush()
+                        slept += 1
+                        total_time_slept += 1
+                        time.sleep(1)
+                    sys.stdout.write("\r                    \n")
+                else:
+                    time.sleep(sleep_time)
+                phase = conn.session.get(phase_url,
+                                              allow_redirects=False).text
+                logging.debug(
+                    "Async transfer Phase for url %s: %s " % (url, phase))
+        except KeyboardInterrupt:
+            # abort the job when receiving a Ctrl-C/Interrupt from the client
+            logging.error("Received keyboard interrupt")
+            conn.session.post(job_url + "/phase",
+                                   allow_redirects=False,
+                                   data="PHASE=ABORT",
+                                   headers={"Content-type": 'text/text'})
+            raise KeyboardInterrupt
+        status = VOFile(phase_url, conn, method="GET",
+                        follow_redirect=False).read().decode('UTF-8')
+
+        logger.debug("Phase:  {0}".format(status))
+        if status in ['COMPLETED']:
+            return False
+        if status in ['HELD', 'SUSPENDED', 'ABORTED']:
+            # re-queue the job and continue to monitor for completion.
+            raise OSError("UWS status: {0}".format(status), errno.EFAULT)
+        error_url = job_url + "/error"
+        error_message = conn.session.get(error_url).text
+        logger.debug(
+            "Got transfer error {0} on URI {1}".format(error_message, uri))
+        # Check if the error was that the link type is unsupported and try and
+        # follow that link.
+        target = re.search('Unsupported link target:(?P<target> .*)$',
+                           error_message)
+        if target is not None:
+            return target.group('target').strip()
+        raise OSError(error_codes.get(error_message, errno.EFAULT),
+                      '{0}: {1}'.format(uri, error_message))
+
+
+class Stream(object):
```

Comment:
Hmm. With this you've opened the can of worms. Maybe it sure ensure the integrity of the data by comparing the source and destination after transfer. And there are the performance issues, which might not be so important for storage. Refer to https://github.com/opencadc/vostools/issues/130
Just wondering if this class should encompass more functionality from vos client and ensure the integrity of the transfer (and possible optimizations)